File: app/src/main/java/com/just_for_fun/dotlist/DBHelper.java
//package com.just_for_fun.dotlist;
//
//import android.content.ContentValues;
//import android.content.Context;
//import android.database.Cursor;
//import android.database.sqlite.SQLiteDatabase;
//import android.database.sqlite.SQLiteOpenHelper;
//import android.util.Log;
//
//import java.util.ArrayList;
//import java.util.List;
//
//// Database Helper Class
//public class DBHelper extends SQLiteOpenHelper {
//
//    // Database Name and Version
//    private static final String DATABASE_NAME = "ToDoDB";
//    private static final int DATABASE_VERSION = 2;
//
//    private static final String COLUMN_NOTES = "notes";
//    private static final String COLUMN_TIMESTAMP = "timestamp";
//    private static final String COLUMN_POSITION = "position";
//
//
//    // Table Name and Columns
//    private static final String TABLE_TASKS = "tasks";
//    private static final String COLUMN_ID = "id";
//    private static final String COLUMN_TITLE = "title";
//    private static final String COLUMN_IS_DONE = "isDone";
//    private static final String COLUMN_FILE_PATH = "filePath";
//
//    private static final String COLUMN_CONTENT = "content";
//    private static final String COLUMN_COMPLETED = "completed";
//
//    // Create Table Query
//    private static final String TABLE_CREATE =
//            "CREATE TABLE " + TABLE_TASKS + " ("
//                    + COLUMN_ID + " INTEGER PRIMARY KEY AUTOINCREMENT, "
//                    + COLUMN_TITLE + " TEXT, "
//                    + COLUMN_CONTENT + " TEXT, "  // Added missing column
//                    + COLUMN_IS_DONE + " INTEGER, "
//                    + COLUMN_COMPLETED + " INTEGER, "  // Added missing column
//                    + COLUMN_FILE_PATH + " TEXT, "
//                    + COLUMN_NOTES + " TEXT, "
//                    + COLUMN_TIMESTAMP + " INTEGER DEFAULT " + System.currentTimeMillis() + ", "
//                    + COLUMN_POSITION + " INTEGER);";
//
//    public DBHelper(Context context) {
//        super(context, DATABASE_NAME, null, DATABASE_VERSION);
//    }
//
//    @Override
//    public void onCreate(SQLiteDatabase db) {
//        db.execSQL(TABLE_CREATE);
//    }
//
//    @Override
//    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
//        if (oldVersion < 2) {
//            db.execSQL("ALTER TABLE " + TABLE_TASKS + " ADD COLUMN " + COLUMN_NOTES + " TEXT");
//            db.execSQL("ALTER TABLE " + TABLE_TASKS + " ADD COLUMN " + COLUMN_TIMESTAMP + " INTEGER DEFAULT " + System.currentTimeMillis());
//            db.execSQL("ALTER TABLE " + TABLE_TASKS + " ADD COLUMN " + COLUMN_POSITION + " INTEGER DEFAULT 0");
//        }
//    }
//
//    // Insert Task
//    public long insertTask(Task task) {
//        SQLiteDatabase db = null;
//        try {
//            db = getWritableDatabase();
//            ContentValues values = new ContentValues();
//
//            // Handle both content-based and title-based tasks
//            if (task.getContent() != null) {
//                values.put(COLUMN_CONTENT, task.getContent());
//            } else {
//                values.put(COLUMN_TITLE, task.getTitle());
//            }
//
//            values.put(COLUMN_IS_DONE, task.isCompleted() ? 1 : 0);
//            values.put(COLUMN_FILE_PATH, task.getFilePath());
//            values.put(COLUMN_POSITION, task.getPosition());
//
//            // Only add notes if details exists
//            if (task.getDetails() != null) {
//                values.put(COLUMN_NOTES, task.getDetails().getNotes());
//            }
//
//            values.put(COLUMN_TIMESTAMP, System.currentTimeMillis());
//
//            return db.insert(TABLE_TASKS, null, values);
//        } finally {
//            if (db != null && db.isOpen()) {
//                db.close();
//            }
//        }
//    }
//
//    // Delete Task
//    public void deleteTask(int id) {
//        SQLiteDatabase db = this.getWritableDatabase();
//        db.delete(TABLE_TASKS, COLUMN_ID + "=?", new String[]{String.valueOf(id)});
//        db.close();
//    }
//
//    public int updateTaskContent(int position, String title) {
//        SQLiteDatabase db = this.getWritableDatabase();
//        ContentValues values = new ContentValues();
//
//        values.put(COLUMN_TITLE, title);
//
//        int rowsAffected = db.update(TABLE_TASKS,
//                values,
//                COLUMN_POSITION + " = ?",
//                new String[]{String.valueOf(position)});
//        db.close();
//        return rowsAffected;
//    }
//
//    // Get all tasks
//    public List<Task> getAllTasks() {
//        List<Task> taskList = new ArrayList<>();
//        String selectQuery = "SELECT * FROM " + TABLE_TASKS + " ORDER BY " + COLUMN_POSITION;
//
//        SQLiteDatabase db = this.getReadableDatabase();
//        Cursor cursor = null;
//
//        try {
//            cursor = db.rawQuery(selectQuery, null);
//
//            if (cursor != null && cursor.moveToFirst()) {
//                int idIndex = cursor.getColumnIndexOrThrow(COLUMN_ID);
//                int titleIndex = cursor.getColumnIndex(COLUMN_TITLE);  // Use getColumnIndex instead
//                int contentIndex = cursor.getColumnIndex(COLUMN_CONTENT);  // Add content index
//                int isDoneIndex = cursor.getColumnIndex(COLUMN_IS_DONE);
//                int filePathIndex = cursor.getColumnIndex(COLUMN_FILE_PATH);
//                int notesIndex = cursor.getColumnIndex(COLUMN_NOTES);
//                int positionIndex = cursor.getColumnIndex(COLUMN_POSITION);
//                int timestampIndex = cursor.getColumnIndex(COLUMN_TIMESTAMP);
//
//                do {
//                    Task task = new Task();
//                    task.setId(cursor.getInt(idIndex));
//
//                    // Handle both content and title based tasks
//                    if (contentIndex != -1 && !cursor.isNull(contentIndex)) {
//                        task.setContent(cursor.getString(contentIndex));
//                    } else if (titleIndex != -1 && !cursor.isNull(titleIndex)) {
//                        task.setTitle(cursor.getString(titleIndex));
//                    }
//
//                    if (isDoneIndex != -1) {
//                        task.setCompleted(cursor.getInt(isDoneIndex) == 1);
//                    }
//
//                    if (filePathIndex != -1) {
//                        task.setFilePath(cursor.getString(filePathIndex));
//                    }
//
//                    if (positionIndex != -1) {
//                        task.setPosition(cursor.getInt(positionIndex));
//                    }
//
//                    TaskDetails details = new TaskDetails();
//                    if (notesIndex != -1) {
//                        details.setNotes(cursor.getString(notesIndex));
//                    }
//                    task.setDetails(details);
//
//                    if (timestampIndex != -1) {
//                        task.setTimestamp(cursor.getLong(timestampIndex));
//                    }
//
//                    taskList.add(task);
//                } while (cursor.moveToNext());
//            }
//        } catch (Exception e) {
//            Log.e("DBHelper", "Error reading tasks: " + e.getMessage());
//        } finally {
//            if (cursor != null) {
//                cursor.close();
//            }
//            db.close();
//        }
//
//        return taskList;
//    }
//
//}

File: app/src/main/java/com/just_for_fun/dotlist/DBHelper.kt
package com.just_for_fun.dotlist

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

// Database Helper Class
class DBHelper(context: Context?) : SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL(TABLE_CREATE)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        if (oldVersion < 2) {
            db.execSQL("ALTER TABLE $TABLE_TASKS ADD COLUMN $COLUMN_NOTES TEXT")
            db.execSQL("ALTER TABLE $TABLE_TASKS ADD COLUMN $COLUMN_TIMESTAMP INTEGER DEFAULT ${System.currentTimeMillis()}")
            db.execSQL("ALTER TABLE $TABLE_TASKS ADD COLUMN $COLUMN_POSITION INTEGER DEFAULT 0")
        }
    }

    // Insert Task
    fun insertTask(task: Task): Long {
        var db: SQLiteDatabase? = null
        try {
            db = writableDatabase
            val values = ContentValues()

            values.put(COLUMN_TITLE, task.title)
            values.put(COLUMN_IS_DONE, if (task.isCompleted) 1 else 0)
            values.put(COLUMN_POSITION, task.position)

            task.details?.let {
                values.put(COLUMN_NOTES, it.notes)
                values.put(COLUMN_FILE_PATH, it.filePath)
            }

            values.put(COLUMN_TIMESTAMP, System.currentTimeMillis())

            val id = db.insert(TABLE_TASKS, null, values)
            return id
        } finally {
            db?.close()
        }
    }

    // Update Task
    fun updateTask(task: Task): Int {
        val db = writableDatabase
        val values = ContentValues().apply {
            put(COLUMN_TITLE, task.title)
            put(COLUMN_IS_DONE, if (task.isCompleted) 1 else 0)
            put(COLUMN_POSITION, task.position)
            put(COLUMN_NOTES, task.details?.notes)
            put(COLUMN_FILE_PATH, task.details?.filePath)
            put(COLUMN_TIMESTAMP, System.currentTimeMillis())
        }

        return db.update(
            TABLE_TASKS,
            values,
            "$COLUMN_POSITION = ?",
            arrayOf(task.position.toString())
        )
    }

    // Delete Task
    fun deleteTask(id: Int) {
        val db = this.writableDatabase
        db.delete(TABLE_TASKS, "$COLUMN_ID=?", arrayOf(id.toString()))
        db.close()
    }

    val allTasks: List<Task>
        get() {
            val taskList: MutableList<Task> = ArrayList()
            val selectQuery = "SELECT * FROM $TABLE_TASKS ORDER BY $COLUMN_POSITION ASC"
            val db = readableDatabase
            var cursor: Cursor? = null

            try {
                cursor = db.rawQuery(selectQuery, null)

                if (cursor?.moveToFirst() == true) {
                    // Required columns
                    val idIndex = cursor.getColumnIndexOrThrow(COLUMN_ID)
                    val positionIndex = cursor.getColumnIndexOrThrow(COLUMN_POSITION)

                    // Optional columns
                    val titleIndex = cursor.getColumnIndex(COLUMN_TITLE)
                    val isDoneIndex = cursor.getColumnIndex(COLUMN_IS_DONE)
                    val notesIndex = cursor.getColumnIndex(COLUMN_NOTES)
                    val filePathIndex = cursor.getColumnIndex(COLUMN_FILE_PATH)
                    val timestampIndex = cursor.getColumnIndex(COLUMN_TIMESTAMP)

                    do {
                        val task = Task(
                            id = cursor.getInt(idIndex),
                            title = if (titleIndex >= 0) cursor.getString(titleIndex) else null,
                            isCompleted = if (isDoneIndex >= 0) cursor.getInt(isDoneIndex) == 1 else false,
                            details = TaskDetails(
                                notes = if (notesIndex >= 0) cursor.getString(notesIndex) else null,
                                filePath = if (filePathIndex >= 0) cursor.getString(filePathIndex) else null
                            ),
                            timestamp = if (timestampIndex >= 0) cursor.getLong(timestampIndex) else System.currentTimeMillis()
                        ).apply {
                            this.position = cursor.getInt(positionIndex)
                        }
                        taskList.add(task)
                    } while (cursor.moveToNext())
                }
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                cursor?.close()
                db.close()
            }

            return taskList
        }


    // Clear All Tasks
    fun clearAllTasks() {
        val db = this.writableDatabase
        db.delete(TABLE_TASKS, null, null)
        db.close()
    }

    fun getTaskByPosition(position: Int): Task? {
        val db = readableDatabase
        var cursor: Cursor? = null
        var task: Task? = null

        try {
            cursor = db.query(
                TABLE_TASKS,
                null,
                "$COLUMN_POSITION = ?",
                arrayOf(position.toString()),
                null,
                null,
                null
            )

            if (cursor?.moveToFirst() == true) {
                // Required columns - use getColumnIndexOrThrow
                val idIndex = cursor.getColumnIndexOrThrow(COLUMN_ID)
                val positionIndex = cursor.getColumnIndexOrThrow(COLUMN_POSITION)

                // Optional columns - use safe handling
                val titleIndex = cursor.getColumnIndex(COLUMN_TITLE)
                val isDoneIndex = cursor.getColumnIndex(COLUMN_IS_DONE)
                val notesIndex = cursor.getColumnIndex(COLUMN_NOTES)
                val filePathIndex = cursor.getColumnIndex(COLUMN_FILE_PATH)
                val timestampIndex = cursor.getColumnIndex(COLUMN_TIMESTAMP)

                task = Task(
                    id = cursor.getInt(idIndex),
                    title = if (titleIndex >= 0) cursor.getString(titleIndex) else null,
                    isCompleted = if (isDoneIndex >= 0) cursor.getInt(isDoneIndex) == 1 else false,
                    details = TaskDetails(
                        notes = if (notesIndex >= 0) cursor.getString(notesIndex) else null,
                        filePath = if (filePathIndex >= 0) cursor.getString(filePathIndex) else null
                    ),
                    timestamp = if (timestampIndex >= 0) cursor.getLong(timestampIndex) else System.currentTimeMillis()
                ).apply {
                    this.position = cursor.getInt(positionIndex)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            return null
        } finally {
            cursor?.close()
            db.close()
        }

        return task
    }

    companion object {
        // Database Name and Version
        private const val DATABASE_NAME = "ToDoDB"
        private const val DATABASE_VERSION = 2

        // Table Name and Columns
        private const val TABLE_TASKS = "tasks"
        private const val COLUMN_ID = "id"
        private const val COLUMN_TITLE = "title"
        private const val COLUMN_CONTENT = "content"
        private const val COLUMN_IS_DONE = "isDone"
        private const val COLUMN_COMPLETED = "completed"
        private const val COLUMN_FILE_PATH = "filePath"
        private const val COLUMN_NOTES = "notes"
        private const val COLUMN_TIMESTAMP = "timestamp"
        private const val COLUMN_POSITION = "position"

        // Create Table Query
        private val TABLE_CREATE = """
            CREATE TABLE $TABLE_TASKS (
                $COLUMN_ID INTEGER PRIMARY KEY AUTOINCREMENT, 
                $COLUMN_TITLE TEXT, 
                $COLUMN_CONTENT TEXT, 
                $COLUMN_IS_DONE INTEGER, 
                $COLUMN_FILE_PATH TEXT,
                $COLUMN_NOTES TEXT, 
                $COLUMN_TIMESTAMP INTEGER DEFAULT ${System.currentTimeMillis()}, 
                $COLUMN_POSITION INTEGER
            );
        """
    }
}
